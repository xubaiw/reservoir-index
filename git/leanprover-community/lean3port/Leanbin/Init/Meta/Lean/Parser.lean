/-
Copyright (c) 2017 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Sebastian Ullrich
-/
prelude
import Leanbin.Init.Meta.Tactic
import Leanbin.Init.Meta.HasReflect
import Leanbin.Init.Control.Alternative

namespace Lean

-- TODO: make inspectable (and pure)
unsafe axiom parser_state : Type

unsafe axiom parser_state.env : parser_state → environment

unsafe axiom parser_state.options : parser_state → options

unsafe axiom parser_state.cur_pos : parser_state → Pos

@[reducible]
unsafe def parser :=
  interaction_monad parser_state

@[reducible]
unsafe def parser_result :=
  interaction_monad.result parser_state

open InteractionMonad

open InteractionMonad.Result

namespace Parser

variable {α : Type}

unsafe def val (p : lean.parser (reflected_value α)) : lean.parser α :=
  reflected_value.val <$> p

protected unsafe class reflectable (p : parser α) where
  full : parser (reflected_value α)

namespace Reflectable

unsafe def expr {p : parser α} (r : reflectable p) : parser expr :=
  reflected_value.expr <$> r.full

unsafe def to_parser {p : parser α} (r : reflectable p) : parser α :=
  val r.full

end Reflectable

unsafe def get_env : parser environment := fun s => success s.env s

unsafe axiom set_env : environment → parser Unit

/-- Make sure the next token is an identifier, consume it, and
    produce the quoted name `t, where t is the identifier. -/
unsafe axiom ident : parser Name

/-- Make sure the next token is a small nat, consume it, and produce it -/
unsafe axiom small_nat : parser Nat

/-- Check that the next token is `tk` and consume it. `tk` must be a registered token. -/
unsafe axiom tk (tk : Stringₓ) : parser Unit

/-- Parse an unelaborated expression using the given right-binding power.
When `pat := tt`, the expression is parsed as a pattern, i.e. local
constants are not checked. -/
protected unsafe axiom pexpr (rbp := Std.Prec.max) (pat := false) : parser pexpr

/-- a variable to local scope -/
unsafe axiom add_local (v : expr) : parser Unit

unsafe axiom add_local_level (v : Name) : parser Unit

unsafe axiom list_include_var_names : parser (List Name)

unsafe axiom list_available_include_vars : parser (List expr)

unsafe axiom include_var : Name → parser Unit

unsafe axiom omit_var : Name → parser Unit

unsafe axiom push_local_scope : parser Unit

unsafe axiom pop_local_scope : parser Unit

/-- Run the parser in a local declaration scope.

Local declarations added via `add_local` do not propagate outside of this scope.
-/
@[inline]
unsafe def with_local_scope {α} (p : parser α) : parser α :=
  interaction_monad.bracket push_local_scope p pop_local_scope

protected unsafe axiom itactic_reflected : parser (reflected_value (tactic Unit))

/-- Parse an interactive tactic block: `begin` .. `end` -/
@[reducible]
protected unsafe def itactic : parser (tactic Unit) :=
  val parser.itactic_reflected

/-- Do not report info from content parsed by `p`. -/
unsafe axiom skip_info (p : parser α) : parser α

/-- Set goal info position of content parsed by `p` to current position. Nested calls take precedence. -/
unsafe axiom set_goal_info_pos (p : parser α) : parser α

/-- Return the current parser position without consuming any input. -/
unsafe def cur_pos : parser Pos := fun s => success (parser_state.cur_pos s) s

/-- Temporarily replace input of the parser state, run `p`, and return remaining input. -/
unsafe axiom with_input (p : parser α) (input : Stringₓ) : parser (α × Stringₓ)

/-- Parse a top-level command. -/
unsafe axiom command_like : parser Unit

unsafe def parser_orelse (p₁ p₂ : parser α) : parser α := fun s =>
  let pos₁ := parser_state.cur_pos s
  result.cases_on (p₁ s) success fun e₁ ref₁ s' =>
    let pos₂ := parser_state.cur_pos s'
    if pos₁ ≠ pos₂ then exception e₁ ref₁ s' else result.cases_on (p₂ s) success exception

unsafe instance : Alternativeₓ parser :=
  { interaction_monad.monad with failure := @interaction_monad.failed _, orelse := @parser_orelse }

-- TODO: move
unsafe def many.{u, v} {f : Type u → Type v} [Monadₓ f] [Alternativeₓ f] {a : Type u} : f a → f (List a)
  | x =>
    (do
        let y ← x
        let ys ← many x
        return <| y :: ys) <|>
      pure List.nil

-- mathport name: «expr ?»
local postfix:100 "?" => optionalₓ

-- mathport name: «expr *»
local postfix:100 "*" => many

unsafe def sep_by : parser Unit → parser α → parser (List α)
  | s, p => List.cons <$> p <*> (s *> p)* <|> return []

unsafe axiom of_tactic : tactic α → parser α

unsafe instance : Coe (tactic α) (parser α) :=
  ⟨of_tactic⟩

namespace Reflectable

unsafe instance cast (p : lean.parser (reflected_value α)) : reflectable (val p) where full := p

unsafe instance has_reflect [r : has_reflect α] (p : lean.parser α) :
    reflectable p where full := do
    let rp ← p
    return ⟨rp⟩

unsafe instance optional {α : Type} [reflected _ α] (p : parser α) [r : reflectable p] :
    reflectable (optionalₓ p) where full := reflected_value.subst some <$> r.full <|> return ⟨none⟩

end Reflectable

unsafe def reflect (p : parser α) [r : reflectable p] : parser expr :=
  r.expr

unsafe axiom run {α} : parser α → tactic α

unsafe def run_with_input {α} : parser α → Stringₓ → tactic α := fun p s => Prod.fst <$> run (with_input p s)

end Parser

end Lean

