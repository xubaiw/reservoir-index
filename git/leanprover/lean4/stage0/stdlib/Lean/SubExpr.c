// Lean compiler output
// Module: Lean.SubExpr
// Imports: Init Lean.Meta.Basic Std.Data.RBMap
#include <lean/lean.h>
#if defined(__clang__)
#pragma clang diagnostic ignored "-Wunused-parameter"
#pragma clang diagnostic ignored "-Wunused-label"
#elif defined(__GNUC__) && !defined(__CLANG__)
#pragma GCC diagnostic ignored "-Wunused-parameter"
#pragma GCC diagnostic ignored "-Wunused-label"
#pragma GCC diagnostic ignored "-Wunused-but-set-variable"
#endif
#ifdef __cplusplus
extern "C" {
#endif
uint64_t lean_uint64_of_nat(lean_object*);
static lean_object* l_Lean_instInhabitedSubExpr___closed__2;
static uint64_t l_Lean_instInhabitedSubExpr___closed__1;
LEAN_EXPORT lean_object* l_Lean_SubExpr_maxChildren;
LEAN_EXPORT lean_object* l_Lean_SubExpr_mkRoot(lean_object*);
static lean_object* l_Lean_instInhabitedSubExpr___closed__3;
LEAN_EXPORT lean_object* l_Lean_instInhabitedSubExpr;
static uint64_t _init_l_Lean_instInhabitedSubExpr___closed__1() {
_start:
{
lean_object* x_1; uint64_t x_2; 
x_1 = lean_unsigned_to_nat(0u);
x_2 = lean_uint64_of_nat(x_1);
return x_2;
}
}
static lean_object* _init_l_Lean_instInhabitedSubExpr___closed__2() {
_start:
{
lean_object* x_1; uint64_t x_2; lean_object* x_3; 
x_1 = lean_unsigned_to_nat(0u);
x_2 = l_Lean_instInhabitedSubExpr___closed__1;
x_3 = lean_alloc_ctor(0, 1, 8);
lean_ctor_set(x_3, 0, x_1);
lean_ctor_set_uint64(x_3, sizeof(void*)*1, x_2);
return x_3;
}
}
static lean_object* _init_l_Lean_instInhabitedSubExpr___closed__3() {
_start:
{
lean_object* x_1; lean_object* x_2; lean_object* x_3; 
x_1 = l_Lean_instInhabitedSubExpr___closed__2;
x_2 = lean_unsigned_to_nat(0u);
x_3 = lean_alloc_ctor(0, 2, 0);
lean_ctor_set(x_3, 0, x_1);
lean_ctor_set(x_3, 1, x_2);
return x_3;
}
}
static lean_object* _init_l_Lean_instInhabitedSubExpr() {
_start:
{
lean_object* x_1; 
x_1 = l_Lean_instInhabitedSubExpr___closed__3;
return x_1;
}
}
static lean_object* _init_l_Lean_SubExpr_maxChildren() {
_start:
{
lean_object* x_1; 
x_1 = lean_unsigned_to_nat(4u);
return x_1;
}
}
LEAN_EXPORT lean_object* l_Lean_SubExpr_mkRoot(lean_object* x_1) {
_start:
{
lean_object* x_2; lean_object* x_3; 
x_2 = lean_unsigned_to_nat(1u);
x_3 = lean_alloc_ctor(0, 2, 0);
lean_ctor_set(x_3, 0, x_1);
lean_ctor_set(x_3, 1, x_2);
return x_3;
}
}
lean_object* initialize_Init(uint8_t builtin, lean_object*);
lean_object* initialize_Lean_Meta_Basic(uint8_t builtin, lean_object*);
lean_object* initialize_Std_Data_RBMap(uint8_t builtin, lean_object*);
static bool _G_initialized = false;
LEAN_EXPORT lean_object* initialize_Lean_SubExpr(uint8_t builtin, lean_object* w) {
lean_object * res;
if (_G_initialized) return lean_io_result_mk_ok(lean_box(0));
_G_initialized = true;
res = initialize_Init(builtin, lean_io_mk_world());
if (lean_io_result_is_error(res)) return res;
lean_dec_ref(res);
res = initialize_Lean_Meta_Basic(builtin, lean_io_mk_world());
if (lean_io_result_is_error(res)) return res;
lean_dec_ref(res);
res = initialize_Std_Data_RBMap(builtin, lean_io_mk_world());
if (lean_io_result_is_error(res)) return res;
lean_dec_ref(res);
l_Lean_instInhabitedSubExpr___closed__1 = _init_l_Lean_instInhabitedSubExpr___closed__1();
l_Lean_instInhabitedSubExpr___closed__2 = _init_l_Lean_instInhabitedSubExpr___closed__2();
lean_mark_persistent(l_Lean_instInhabitedSubExpr___closed__2);
l_Lean_instInhabitedSubExpr___closed__3 = _init_l_Lean_instInhabitedSubExpr___closed__3();
lean_mark_persistent(l_Lean_instInhabitedSubExpr___closed__3);
l_Lean_instInhabitedSubExpr = _init_l_Lean_instInhabitedSubExpr();
lean_mark_persistent(l_Lean_instInhabitedSubExpr);
l_Lean_SubExpr_maxChildren = _init_l_Lean_SubExpr_maxChildren();
lean_mark_persistent(l_Lean_SubExpr_maxChildren);
return lean_io_result_mk_ok(lean_box(0));
}
#ifdef __cplusplus
}
#endif
