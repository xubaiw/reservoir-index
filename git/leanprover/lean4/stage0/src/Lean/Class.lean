/-
Copyright (c) 2019 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
import Lean.Attributes

namespace Lean

structure ClassEntry where
  name      : Name
  /--
    Position of the class `outParams`.
    For example, for class
    ```
    class GetElem (Cont : Type u) (Idx : Type v) (Elem : outParam (Type w)) (Dom : outParam (Cont → Idx → Prop)) where
    ```
    `outParams := #[2, 3]`
  -/
  outParams : Array Nat

namespace ClassEntry

def lt (a b : ClassEntry) : Bool :=
  Name.quickLt a.name b.name

end ClassEntry

structure ClassState where
  outParamMap : SMap Name (Array Nat) := SMap.empty
  deriving Inhabited

namespace ClassState

def addEntry (s : ClassState) (entry : ClassEntry) : ClassState :=
  { s with outParamMap := s.outParamMap.insert entry.name entry.outParams }

def switch (s : ClassState) : ClassState :=
  { s with outParamMap := s.outParamMap.switch }

end ClassState

/- TODO: add support for scoped instances -/
builtin_initialize classExtension : SimplePersistentEnvExtension ClassEntry ClassState ←
  registerSimplePersistentEnvExtension {
    name          := `classExt
    addEntryFn    := ClassState.addEntry
    addImportedFn := fun es => (mkStateFromImportedEntries ClassState.addEntry {} es).switch
  }

@[export lean_is_class]
def isClass (env : Environment) (n : Name) : Bool :=
  (classExtension.getState env).outParamMap.contains n

/--
  If `declName` is a class, return the position of its `outParams`.
-/
def getOutParamPositions? (env : Environment) (declName : Name) : Option (Array Nat) :=
  (classExtension.getState env).outParamMap.find? declName

@[export lean_has_out_params]
def hasOutParams (env : Environment) (declName : Name) : Bool :=
  match getOutParamPositions? env declName with
  | some outParams => !outParams.isEmpty
  | none => false

/--
  Auxiliary function for collection the position class `outParams`, and
  checking whether they are being correctly used.
  A regular (i.e., non `outParam`) must not depend on an `outParam`.
  Reason for this restriction:
  When performing type class resolution, we replace arguments that
  are `outParam`s with fresh metavariables. If regular parameters could
  depend on `outParam`s, then we would also have to replace them with
  fresh metavariables. Otherwise, the resulting expression could be type
  incorrect. This transformation would be counterintuitive to users since
  we would implicitly treat these regular parameters as `outParam`s.
-/
private partial def checkOutParam (i : Nat) (outParamFVarIds : Array FVarId) (outParams : Array Nat) (type : Expr) : Except String (Array Nat) :=
  match type with
  | .forallE _ d b _ =>
    if d.isOutParam then
      let fvarId := { name := Name.mkNum `_fvar outParamFVarIds.size }
      let fvar      := mkFVar fvarId
      let b         := b.instantiate1 fvar
      checkOutParam (i+1) (outParamFVarIds.push fvarId) (outParams.push i) b
    else if d.hasAnyFVar fun fvarId => outParamFVarIds.contains fvarId then
      Except.error s!"invalid class, parameter #{i+1} depends on `outParam`, but it is not an `outParam`"
    else
      checkOutParam (i+1) outParamFVarIds outParams b
  | _ => return outParams

def addClass (env : Environment) (clsName : Name) : Except String Environment := do
  if isClass env clsName then
    throw s!"class has already been declared '{clsName}'"
  let some decl := env.find? clsName
    | throw s!"unknown declaration '{clsName}'"
  unless decl matches .inductInfo .. | .axiomInfo .. do
    throw s!"invalid 'class', declaration '{clsName}' must be inductive datatype, structure, or constant"
  let outParams ← checkOutParam 0 #[] #[] decl.type
  return classExtension.addEntry env { name := clsName, outParams }

private def consumeNLambdas : Nat → Expr → Option Expr
  | 0,   e            => some e
  | i+1, .lam _ _ b _ => consumeNLambdas i b
  | _,   _            => none

partial def getClassName (env : Environment) : Expr → Option Name
  | .forallE _ _ b _ => getClassName env b
  | e                => do
    let Expr.const c _ ← pure e.getAppFn | none
    let info ← env.find? c
    match info.value? with
    | some val => do
      let body ← consumeNLambdas e.getAppNumArgs val
      getClassName env body
    | none =>
      if isClass env c then some c
      else none

builtin_initialize
  registerBuiltinAttribute {
    name  := `class
    descr := "type class"
    add   := fun decl stx kind => do
      let env ← getEnv
      Attribute.Builtin.ensureNoArgs stx
      unless kind == AttributeKind.global do throwError "invalid attribute 'class', must be global"
      let env ← ofExcept (addClass env decl)
      setEnv env
  }

end Lean
