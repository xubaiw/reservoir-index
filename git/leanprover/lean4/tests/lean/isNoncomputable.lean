import Lean
open Lean

def test (declName : Name) : MetaM Unit := do
  if isNoncomputable (‚Üê getEnv) declName then
    IO.println s!"{declName} is marked as noncomputable"
  else
    IO.println s!"{declName} is not marked as noncomputable"

noncomputable def foo (x : Nat) : Nat :=
  x + Classical.ofNonempty

#eval test ``List.map
#eval test ``foo
#eval test ``Classical.ofNonempty
#eval test ``Array.map
