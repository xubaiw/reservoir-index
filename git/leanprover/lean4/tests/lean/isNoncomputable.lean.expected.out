List.map is not marked as noncomputable
foo is marked as noncomputable
Classical.ofNonempty is marked as noncomputable
Array.map is not marked as noncomputable
